File: Mobiles/flutter/lib/core/application_layer/o11y/errors/o11y_errors.dart
import 'package:faro/faro.dart';
import 'package:flutter_mobile_o11y_demo/core/application_layer/o11y/faro/faro.dart';

class O11yErrors {
  O11yErrors() : _faro = faro;

  final Faro _faro;

  void reportError({
    required String type,
    required String error,
    StackTrace? stacktrace,
    Map<String, String>? context,
  }) {
    _faro.pushError(
      type: type,
      value: error,
      stacktrace: stacktrace,
      context: context,
    );
  }
}

final o11yErrors = O11yErrors();

File: Mobiles/react-native/src/core/o11y/o11yErrors.ts
import { faro } from '@grafana/faro-react-native';

export function reportError(options: {
  type: string;
  error: string;
  stacktrace?: string;
  context?: Record<string, string>;
}): void {
  if (!faro?.api) return;

  const context: Record<string, string> = { ...options.context };
  const err = new Error(options.error);
  /**
   * Faro builds `stacktrace.frames` from `error.stack` via `parseStacktrace`.
   * Assigning the caught stack here ensures handled exceptions use the real Hermes positions
   * for symbolification — not the shallow stack from `new Error()` inside this helper.
   */
  if (options.stacktrace) {
    err.stack = options.stacktrace;
  }

  faro.api.pushError(err, {
    type: options.type,
    context,
  });
}
